### DISCUSSION TITLE
关于蒙哥马利简约实现

### DISCUSSION THREAD

**Discussion by kingstenzzz** [emmansun/gmsm#3] [Q&A] [answered]
想问一下这里面的实现细节，
1.这里的position指的是哪里？
https://github.com/emmansun/gmsm/blob/b66f1b467cc7240dc361a5cd7638688082bb6265/sm2/p256.go#L568
2.这个  have a factor of 0xffffc00 = 2^29 - 2^11的意思是什么？https://github.com/emmansun/gmsm/blob/b66f1b467cc7240dc361a5cd7638688082bb6265/sm2/p256.go#L613

3./ Word:                2   3   4   5   6   7   8   9   10
		// Added in top half:   29 28  29  29  29  29  29  28
		//                             29  28  29  28  29
		//                                             29
		// Added in bottom half:   28  29  28  28  28  29  28   28
		//                                 28  29  28  29  28
这个表得意思不是很理解。

**Comment by emmansun** [comment, marked as answer]
我从第三个问题开始回答：

三，首先这个除R的约简是通过不断加上tmp2数组最右（低）字（word）乘P来达成的，一次循环乘两次，虽然都是乘P，但是top half乘的是29 28 29 28 ... 这样位宽表示的P `{0x1fffffff, 0xfffffff, 0x7f, 0xffffc00, 0x1fffffff, 0xfffffff, 0x1fffffff, 0xeffffff, 0xfffffff}`，而bottom half是乘的28 29 28 29... 这样位宽表示的P `{0xfffffff, 0x1fffffff, 0x7f, 0x1ffff800, 0xfffffff, 0x1fffffff, 0xfffffff, 0x1dffffff, 0xfffffff}`，之所以这样是为了和tmp2中的每个word的位宽保持一致，这样方便把乘法表示成加减和位移运算。

二， 就是x 要乘以 0x1ffff800，而0x1ffff800表示成2^29 - 2^11，这是bottom half中乘以P的第四个word。

一，这里的position是指的bit表示的起始位置，参考方法开始部分的注释。
![image](https://user-images.githubusercontent.com/7235232/131619957-8da991d0-ce34-4089-ab7f-66c1a971f3bb.png)

**Comment by kingstenzzz** [comment]
> 我从第三个问题开始回答：
> 
> 三，首先这个除R的约简是通过不断加上tmp2数组最右（低）字（word）乘P来达成的，一次循环乘两次，虽然都是乘P，但是top half乘的是29 28 29 28 ... 这样位宽表示的P `{0x1fffffff, 0xfffffff, 0x7f, 0xffffc00, 0x1fffffff, 0xfffffff, 0x1fffffff, 0xeffffff, 0xfffffff}`，而bottom half是乘的28 29 28 29... 这样位宽表示的P `{0xfffffff, 0x1fffffff, 0x7f, 0x1ffff800, 0xfffffff, 0x1fffffff, 0xfffffff, 0x1dffffff, 0xfffffff}`，之所以这样是为了和tmp2中的每个word的位宽保持一致，这样方便把乘法表示成加减和位移运算。
> 
> 二， 就是x 要乘以 0x1ffff800，而0x1ffff800表示成2^29 - 2^11，这是bottom half中乘以P的第四个word。
> 
> 一，这里的position是指的bit表示的起始位置，参考方法开始部分的注释。
> ![image](https://user-images.githubusercontent.com/7235232/131619957-8da991d0-ce34-4089-ab7f-66c1a971f3bb.png)

感谢回复，基本上原理是懂得，但是一些实现的细节不是很清楚想问一下，方便留下一个联系方式（邮箱或者qq）吗？

**Comment by emmansun** [comment]
我已在本项目中开放discussions功能。
